src/perform_step/bdf_perform_step.jl

```diff
@@ -262,19 +256,200 @@ function perform_step!(integrator,cache::QNDF1Cache,repeat_step=false)
   fail_convergence && return
   @. u = tmp + γ*z
 
-  if integrator.opts.adaptive
-    D2!(u,uprev,k,cache)
-    @. utilde = (κ*γₖ + inv(k+1)) * D2[1,1]
+  if integrator.opts.adaptive && integrator.success_iter > 0
+    @. D2[1] = u - uprev
+    @. D2[2] = D2[1] - D[1]
+    @. utilde = (κ*γ₁ + inv(k+1)) * D2[2]
     calculate_residuals!(atmp, utilde, uprev, u, integrator.opts.abstol, integrator.opts.reltol, integrator.opts.internalnorm)
     integrator.EEst = integrator.opts.internalnorm(atmp)
     if integrator.EEst > one(integrator.EEst)
-      cache.D .= temp_D
-      cache.uprev2 .= temp_u
       return
     end
+  else
+    integrator.EEst = one(integrator.EEst)
+  end
+  cache.dtₙ₋₁ = dt
+  cache.uprev2 .= uprev
+  cache.ηold = η
+  cache.newton_iters = iter
+  f(integrator.fsallast, u, p, t+dt)
+end
+
+function initialize!(integrator, cache::QNDF2ConstantCache)
+  integrator.kshortsize = 2
+  integrator.k = typeof(integrator.k)(integrator.kshortsize)
+  integrator.fsalfirst = integrator.f(integrator.uprev, integrator.p, integrator.t) # Pre-start fsal
+
+  # Avoid undefined entries if k is an array of arrays
+  integrator.fsallast = zero(integrator.fsalfirst)
+  integrator.k[1] = integrator.fsalfirst
+  integrator.k[2] = integrator.fsallast
+end
+
+function perform_step!(integrator,cache::QNDF2ConstantCache,repeat_step=false)
+  @unpack t,dt,uprev,u,f,p = integrator
+  @unpack uprev2,uprev3,dtₙ₋₁,dtₙ₋₂,D,D2,R,U = cache
+  cnt = integrator.iter
+  k = 2
+  if cnt == 1 || cnt == 2
```

Comment:
why if cnt == 2?

Reply:
oh, I forgot to remove.

Reply:
oh, It is Okay after two-step we can apply qndf2, remember it has an extra term `kappa*gamma_k*(y_{n+1}-y_{n+1}^0))`.